keras/caffe/converters.py

```diff
@@ -0,0 +1,399 @@
+
+	starts_names = []
+	for start in starts:
+		starts_names.append(layers[start].name)
+	ends_names = []
+	for end in ends:
+		ends_names.append(layers[end].name)
+
+	return model, starts_names, ends_names
+
+def model_from_param(layers):
+	'''
+		layers: a list of all the layers in the model with the assocaited parameters
+	'''
+	# DEPLOY MODE: first layer is computation
+	# NON DEPLOY MODE: first layer is data input
+
+	network = make_network(layers, 0)	# obtain the nodes that make up the graph
+	if len(network) == 0:
+		raise Exception('failed to construct network from the caffemodel')
+	network = acyclic(network)	# Convert it to be truly acyclic
+	network = merge_layer_blob(network)	# eliminate 'blobs', just have layers
+	reverse_network = reverse(network)	# reverse, to obtain the start
+	starts = get_starts(reverse_network)	# inputs of the network - 'in-order' is zero
+	ends = get_ends(network)	# outputs of the network - 'out-order' is zero
+	network = remove_label_paths(network, starts, ends)	# path from input to loss layers removed
+	reverse_network = reverse(network) # stores the 'input' for each layer
+
+	model = Graph()
+
+	# add input nodes
+	for start in starts:
+		name = layers[start].name
+		model.add_input(name=name, ndim=4)
+
+	# parse all the layers and build equivalent keras graph
+	for layer_nb in network:
+		layer = layers[layer_nb]
+		name = layer.name
+
+		if is_data_input(layer):
+			continue
+
+		if layer_nb in starts:
+			input_layer_names = [name]	# this is both an input layer and computation layer. the name has been added as input layer
+			name = 'graph_input_' + name # use a prefix to create a new name, which is used on from hereon to take input from this layer(see below)
+		else:
+			input_layers = reverse_network[layer_nb]	# inputs to current layer, in the form of layer numbers
+			input_layer_names = []	# list of strings identifying the input layer
+			for input_layer in input_layers:
+				if input_layer in starts and not is_data_input(layers[input_layer]):	# taking input from first layer and that's not a data layer, implies use the prefix 
+					input_layer_names.append('graph_input_' + layers[input_layer].name)
+				else:
+					input_layer_names.append(layers[input_layer].name)
+
+		if len(input_layer_names) == 1:
+			input_layer_name = input_layer_names[0] # single input. since concatenation is explicit, 
+													# all layers can be thought of single input layers 
+													# (except loss layers, which is handled anyway)
+		if layer_nb in ends:
+			name = 'output_' + name 	# outputs nodes are marked with 'output_' prefix from which output is derived
+										# since the graph output is not model output
+
+
+		if layer.type == 3:
+			# CONCAT
+			# emulation of just concatenation
+			axis = layer.concat_param.axis # 0 for batch, 1 for stack
+			model.add_node(ZeroPadding2D(pad=(0, 0)), name=name, inputs=input_layer_names, concat_axis=axis)
```

Comment:
What's the `ZeroPadding2D(pad=(0, 0))` for? Doesn't it do nothing? If you need a layer that does nothing, `Activation('linear')` would be more efficient and straightforward.